fix!: payref migration and indexes, add grpc query via output hash

hansieodendaal opened the pull request:
Description
---
- Added gRPC method to retrieve output information via output hash
- Updated the gRPC payref search to return spent and unspent output information
- Added migration to rebuild payref indexes due to missing payrefs (_added periodic call to `LMDBStore::resize_if_required`_)
- Fixed error in lmdb migration counter where it recorded a higher version done than what it should

Fixes #7263 

Motivation and Context
---
- Payref information was deleted when the outputs were spent.
- With migrations, payref information was only created for the current output set, not all outputs.
- Payref lookups for spent outputs were not possible.
- Output and payref information was not possible via output hash.

How Has This Been Tested?
---
- New lmdb migration tested at the system level for fresh and existing base nodes.
- New gRPC method system-level testing [_tested with block explorer upgrade_].
- General system-level testing.

Fresh base node with a restart afterwards
```rust
2025-06-27 09:09:24.013840800 [c::cs::lmdb_db::lmdb_db] INFO  [MIGRATIONS] Blockchain database is at v0 (required version: 5)
2025-06-27 09:09:24.013873800 [c::cs::lmdb_db::lmdb_db] INFO  [MIGRATIONS] v1: No accumulated difficulty found for block height 14999
2025-06-27 09:09:24.013877800 [c::cs::lmdb_db::lmdb_db] INFO  [MIGRATIONS] v1: No migration to perform for version network
2025-06-27 09:09:24.013883100 [c::cs::lmdb_db::lmdb_db] INFO  [MIGRATIONS] v2: Starting PayRef migration
2025-06-27 09:09:24.013899700 [c::cs::lmdb_db::lmdb_db] INFO  [MIGRATIONS] v2: Cleared PayRef index
2025-06-27 09:09:24.038671500 [c::cs::lmdb_db::lmdb_db] INFO  [MIGRATIONS] v2: PayRef index rebuild completed
2025-06-27 09:09:24.038693200 [c::cs::lmdb_db::lmdb_db] INFO  [MIGRATIONS] Migrated database from version 2 to version 3
2025-06-27 09:09:24.039646900 [c::cs::lmdb_db::lmdb_db] INFO  [MIGRATIONS] Migrated database from version 3 to version 4
2025-06-27 09:09:24.040506700 [c::cs::lmdb_db::lmdb_db] INFO  [MIGRATIONS] Migrated database from version 4 to version 5
2025-06-27 09:10:24.526178200 [c::cs::lmdb_db::lmdb_db] INFO  [MIGRATIONS] Blockchain database is at v5 (required version: 5)
```

What process can a PR reviewer use to test or verify this change?
---
- Code review.
- System-level testing.

<!-- Checklist -->
<!-- 1. Is the title of your PR in the form that would make nice release notes? The title, excluding the conventional commit
tag, will be included exactly as is in the CHANGELOG, so please think about it carefully. -->


Breaking Changes
---

- [ ] None
- [ ] Requires data directory on base node to be deleted
- [ ] Requires hard fork
- [X] Other - Please specify

<!-- Does this include a breaking change? If so, include this line as a footer -->
BREAKING CHANGE: 
- A new base node gRPC method `SearchPaymentReferencesViaOutputHash` was added that returns a stream of `PaymentReferenceResponse`.
- The gRPC `PaymentReferenceResponse` was also updated to return the spent timestamp and output hash additionally.


<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

## Summary by CodeRabbit

* **New Features**
  * Added a new gRPC method to search payment references by output hash, providing detailed mined and spent information for outputs.
  * Expanded response data to include spent timestamps and output hashes for enhanced transaction tracing.

* **Refactor**
  * Unified and extended backend logic for fetching mined information by payment reference or output hash.
  * Improved migration and index handling for payment references to ensure data consistency, including genesis block coverage.

* **Documentation**
  * Clarified field comments and updated configuration files to reflect new gRPC capabilities.

* **Style**
  * Updated test and code comments for clarity and consistency.
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

Commit:
Add payref reverse lookup

- Added payref reverse lookup lmdb table
- Added migration to rebuild payref indexes and to add reverse lookup lmdb table
- Fixed error in lmdb migration counter
- Added gRPC method to retrieve output information via output hash
- Updated the gRPC payref search to return spent and unspent output information

Commit:
Merge remote-tracking branch 'upstream/development' into ho_outputs_by_hash

hansieodendaal commented on base_layer/core/src/chain_storage/blockchain_database.rs:
```diff
@@ -454,10 +454,22 @@ where B: BlockchainBackend
         let db = self.db_read_access()?;
-        db.fetch_output_by_payref(&payref)
+        db.fetch_mined_info_by_payref(&payref)
+    }
```
See `MinedInfo` below - it consists of options
```
pub struct MinedInfo {
    pub input: Option<InputMinedInfo>,
    pub output: Option<OutputMinedInfo>,
}
```

Commit:
fix unit test issue, better logging

SWvheerden reviewed (changes requested):
I am against this appraoch
We already have all info needed to calculate this, we don't need to store another index, 
And on the application layer, you need to do two lookups and not one. 

use this function: `fetch_output_in_txn`
it contains all the info needed for the payref.

SWvheerden commented on base_layer/core/src/chain_storage/lmdb_db/lmdb_db.rs:
```diff
@@ -3150,41 +3222,48 @@ fn run_migrations(db: &mut LMDBDatabase) -> Result<(), ChainStorageError> {
                 &db.metadata_db,
                 &k.as_u32(),
-                &MetadataValue::MigrationVersion(migrate_from_version + 1),
+                &MetadataValue::MigrationVersion(max(migrate_from_version + 1, MIGRATION_VERSION)),
```
```suggestion
                &MetadataValue::MigrationVersion(max(migrate_from_version , MIGRATION_VERSION)),
```
thats why the migration ended in a 4 and not the correct 3, we can just skip 4 and

hansieodendaal replied:
I solved this in a slightly differtent manner - system-level tested and working

Commit:
fix payref migration

Commit:
review comments, fix issues

Commit:
Merge remote-tracking branch 'upstream/development' into ho_outputs_by_hash

Commit:
Fix critical db needs resize error

[c::cs::lmdb_db::lmdb_db] TRACE Processing payref 668624580bc1d2675de974c444c9ec0b1052bf41f14d673defa01fccfcc36175 and output hash 1a11b329183df05848249ee345ee29e44375505dac1689adc32ac021d0f636f6 for height 26902
[c::cs::lmdb_db::lmdb_db] TRACE Processing payref 5cac218ded22eeb8efdc4dc295a2461666c8d5137abfc02b30b9152f82332e99 and output hash 1a5551e00b270645e3804be945d477840a748bca9f96dd919a3691271c1bef43 for height 26902
[c::cs::lmdb_db::lmdb_db] TRACE Processing payref ae08d3ec581531f503603f419ff458e28eea07669286f7d3c650a0b6bcbc0f40 and output hash 1c0a293ed8682b4f00bfa9d5cb8ee039f1065a8aaff964c91318bfe8e910316f for height 26902
[tonic::transport::server] TRACE signal received, shutting down
[tonic::transport::server] TRACE waiting for 0 connections to close
[minotari::base_node::app] INFO  Stopping GRPC
[minotari::base_node::app] ERROR Exiting with code (114): ExitError { exit_code: DatabaseError, details: Some("Database resize required") }

Commit:
db resize during payref migration

SWvheerden commented on base_layer/core/src/chain_storage/lmdb_db/lmdb_db.rs:
```diff
@@ -608,13 +611,7 @@ impl LMDBDatabase {
-            &output_hash,
-            "payref_to_output_index",
-        )?;
+        lmdb_replace(txn, &self.payref_to_output_index, payref.as_slice(), &output_hash, None)?;
```
```suggestion
        lmdb_insert(txn, &self.payref_to_output_index, payref.as_slice(), &output_hash, None)?;
```
this will hide errors

hansieodendaal replied:
Agreed, this was due to double processing the genesis block, once in the migration and then when building the chain.

SWvheerden commented on base_layer/core/src/chain_storage/lmdb_db/lmdb_db.rs:
```diff
@@ -3117,10 +3112,16 @@ fn run_migrations(db: &mut LMDBDatabase) -> Result<(), ChainStorageError> {
             txn.commit()?;
         }
+
+        // MIGRATION: Accumulated difficulty migration - blockchain rewind to last known good accumulated difficulty
```
```suggestion
        // MIGRATION: Total Accumulated difficulty migration - blockchain rewind to last known good accumulated difficulty
```

SWvheerden commented on base_layer/core/src/chain_storage/lmdb_db/lmdb_db.rs:
```diff
@@ -3132,59 +3133,78 @@ fn run_migrations(db: &mut LMDBDatabase) -> Result<(), ChainStorageError> {
+
+        // Lets update the migration version
         {
+            let migrated_to_version = migrate_from_version + 1;
```
```suggestion
            let migrated_to_version = migrate_from_version ;
```
was the +1 not a problem

hansieodendaal replied:
No, it was the combination of the `migrate_from_version` counter, where it ended and what is deemed as the version we have achieved - to make this clear I introduced the `migrated_to_version` variable and changed `last_migrated_version..=MIGRATION_VERSION` to `last_migrated_version..MIGRATION_VERSION`. 

If we use `last_migrated_version..=MIGRATION_VERSION` the start and end logic is not 100%  as it would revisit the loop if we are on v4 and we should be on v4 - now it will only go into the loop if a migration is needed.

SWvheerden commented on base_layer/core/src/chain_storage/lmdb_db/lmdb_db.rs:
```diff
@@ -3132,59 +3133,78 @@ fn run_migrations(db: &mut LMDBDatabase) -> Result<(), ChainStorageError> {
             drop(read_txn);
-            // we need to clear the payref index first as they might now bw wrong
+
+            // Clear the payref index first as they might now be wrong
```
we dont need to delete it

hansieodendaal replied:
Agreed, a remnant of the previous code - replace will update it anyway.

SWvheerden commented on base_layer/core/src/chain_storage/lmdb_db/lmdb_db.rs:
```diff
@@ -3287,33 +3307,62 @@ fn get_correct_accumulated_difficulty() -> Vec<(u64, U512)> {
+    debug!(target: LOG_TARGET, "Processing PayRef migration for {}", height);
+
+    let (block_hash, output_row_data) = if height == 0 {
+        let genesis_block = get_genesis_block(Network::get_current_or_user_setting_or_default());
+        let header_hash = genesis_block.header().hash();
+        let mined_timestamp = genesis_block.header().timestamp.as_u64();
+        let outputs = genesis_block
+            .block()
+            .body
+            .outputs()
+            .iter()
+            .map(|o| TransactionOutputRowData {
+                output: o.clone(),
+                header_hash,
+                hash: o.hash(),
+                mined_height: 0,
+                mined_timestamp,
+            })
+            .collect::<Vec<_>>();
+        (header_hash, outputs)
```
this does not seem correct, you are double-processing the block here. 
If the block chain is empty, this is going to be 0 by the unwrap or 0 above, which means the gen block still needs to be processed. If its actually 0, it should be in the outputs, and we can recall them from that. 
Is this migration the reason you changed the lmdb_insert to lmdb_replace?

hansieodendaal replied:
Yes, that is the issue, double processing, and that was the reason for the replace

Commit:
review comments

SWvheerden commented on base_layer/core/src/chain_storage/lmdb_db/lmdb_db.rs:
```diff
@@ -1084,16 +1057,15 @@ impl LMDBDatabase {
-
             // Clean up PayRef index for this output during reorg
             // Only clean PayRef if output was actually created (not spent in same block, not burned)
             let should_cleanup_payref =
```
this should be updated as well

SWvheerden commented on base_layer/core/src/chain_storage/lmdb_db/lmdb_db.rs:
```diff
@@ -3164,27 +3182,33 @@ fn run_migrations(db: &mut LMDBDatabase) -> Result<(), ChainStorageError> {
             for height in 0..=chain_height {
+                if height % 100 == 0 {
+                    unsafe {
+                        LMDBStore::resize_if_required(&db.env, &db.env_config, None)?;
```
I think this is somewhat in the correct order, but call this between all migrations, 
And make this: every X block, 100 is too few...
a payref index entry is 32*2 bytes, default increase is 128mb.

SWvheerden commented on applications/minotari_app_grpc/proto/base_node.proto:
```diff
@@ -591,20 +594,24 @@ message SearchPaymentReferencesRequest {
+    // Block hash where output was spent - will be empty for unspent outputs
     bytes spent_block_hash = 8;
-    // Transaction output amount will be 0 for non set a
+    // Minimum value promise - will be 0 for spent outputs
```
I dont think this comment is correct?

hansieodendaal replied:
We populate it with 0 for spent outputs as we do not know what it is

SWvheerden reviewed (changes requested):
Looking good, 
There is a missed if statement that needs updating, and I think we need to update the migration thing as noted.

Commit:
review comments

MCozhusheck commented:
Code is looking great, tested manually and search by output hash `99fa77666a4c06bbdaebf3c9fad441e1cd82116df16e9754a7fae57b229a34a0` (first output in block 15756) and found exact payref https://explore.tari.com/blocks/15756
Tested for multiple inputs and invalid inputs. In a single input that is invalid I got error response but in case of multiple inputs where some were invalid I got response only for valid one.

hansieodendaal commented on applications/minotari_app_grpc/proto/base_node.proto:
```diff
@@ -591,20 +594,24 @@ message SearchPaymentReferencesRequest {
+    // Block hash where output was spent - will be empty for unspent outputs
     bytes spent_block_hash = 8;
-    // Transaction output amount will be 0 for non set a
+    // Minimum value promise - will be 0 for spent outputs as we do not track this for spent outputs
```
```suggestion
    // Minimum value promise
```

hansieodendaal committed:
Update applications/minotari_app_grpc/proto/base_node.proto

SWvheerden commented on applications/minotari_app_grpc/proto/base_node.proto:
```diff
@@ -591,20 +594,24 @@ message SearchPaymentReferencesRequest {
     // The payment reference that was found
     string payment_reference_hex = 1;
-    // Block height where the output was mined
+    // Block height where the output was mined - will be 0 for spent outputs as it is not applicable
```
and these all?

hansieodendaal replied:
```suggestion
    // Block height where the output was mined
```

hansieodendaal commented on applications/minotari_app_grpc/proto/base_node.proto:
```diff
@@ -591,20 +594,24 @@ message SearchPaymentReferencesRequest {
+    // Block height where the output was mined - will be 0 for spent outputs as it is not applicable
     uint64 block_height = 2;
-    // Block hash where the output was mined
+    // Block hash where the output was mined - will be empty for spent outputs as it is not applicable
```
```suggestion
    // Block hash where the output was mined
```

hansieodendaal commented on applications/minotari_app_grpc/proto/base_node.proto:
```diff
@@ -591,20 +594,24 @@ message SearchPaymentReferencesRequest {
+    // Block hash where the output was mined - will be empty for spent outputs as it is not applicable
     bytes block_hash = 3;
-    // Timestamp when the output was mined
+    // Timestamp when the output was mined - will be 0 for spent outputs as it is not applicable
```
```suggestion
    // Timestamp when the output was mined
```

hansieodendaal committed:
Update applications/minotari_app_grpc/proto/base_node.proto

hansieodendaal committed:
Update applications/minotari_app_grpc/proto/base_node.proto

hansieodendaal committed:
Update applications/minotari_app_grpc/proto/base_node.proto

hansieodendaal commented on applications/minotari_app_grpc/proto/base_node.proto:
```diff
@@ -597,14 +600,18 @@ message PaymentReferenceResponse {
     // Timestamp when the output was mined
     uint64 mined_timestamp = 4;
-    // Output commitment (32 bytes)
+    // Output commitment (32 bytes) - will be empty if output was spent as we do not track this for spent outputs
```
```suggestion
    // Output commitment (32 bytes)
```

hansieodendaal committed:
Update applications/minotari_app_grpc/proto/base_node.proto
